### PR TITLE
Avoid X11 dependency using an environment variable.

### PR DESCRIPTION
**Pull request by set-soft**
This patch delays the wx.App() initialization only if the INTERACTIVE_HTML_NO_X11 is defined to "True". Otherwise the initialization is done, even if no dialog is used.

See INTI-CMNB/InteractiveHtmlBom@058ade94eb2934c4e1ec1d9d8ab54ff51d937f5b

### CODE REVIEW

File: InteractiveHtmlBom/generate_interactive_bom.py [hunk 1]
```diff
@@ -29,7 +27,13 @@ def to_utf(s):
     from .ecad import get_parser_by_extension
     from .version import version
 
-    app = wx.App()
+    delay_wx = False
+    if os.environ.get('INTERACTIVE_HTML_NO_X11') == 'True':
```
Comment (qu1ck) [inline InteractiveHtmlBom/generate_interactive_bom.py:31]: X11 is hopefully going to the software heaven soon and will be replaced by modern alternatives (read wayland).
`INTERACTIVE_HTML_BOM_NO_DISPLAY` is more generic.

Also for such flag variables usually just their existence is checked, not specific value. Also helps with case variations and representations of boolean. 

Another thing is the semantics of this variable. It's purpose is to avoid dependency on display server and yet the logic doesn't avoid the dependency, just delays it. It makes no sense to ever set this variable and use `--show-dialog` because that flag implies that you have a display to show that dialog on. I think it's best to just not create wxApp at all if the environment variable is set and exit with error if `--show-dialog` is specified at the same time.

I suggest this fix:
```python
create_wx_app = 'INTERACTIVE_HTML_BOM_NO_DISPLAY' not in os.environ
if create_wx_app:
  import wx
  app = wx.App()
```

File: InteractiveHtmlBom/generate_interactive_bom.py [hunk 2]
```diff
@@ -47,6 +51,9 @@ def to_utf(s):
     logger = ibom.Logger(cli=True)
     parser = get_parser_by_extension(os.path.abspath(args.file), config, logger)
     if args.show_dialog:
+        if delay_wx:
+            import wx
+            app = wx.App()
```
Comment (qu1ck) [inline InteractiveHtmlBom/generate_interactive_bom.py:56]: Similarly here:
```python
if not create_wx_app:
  print("Can not show dialog when INTERACTIVE_HTML_BOM_NO_DISPLAY is set.")
  exit(1)
```

Comment (set-soft) [reply InteractiveHtmlBom/generate_interactive_bom.py:56]: Ok. I agree this is more coherent.

**Comment by set-soft** [comment]
I changed it to be even simpler, just detecting if DISPLAY is defined.
If DISPLAY is not defined or empty the code avoids creating wx.App, which we know will fail.

**Comment by qu1ck** [comment]
Checking for Ibom specific variable is preferable because `$DISPLAY` is linux only thing.

Also I'd rather have users go through extra step of defining a variable to disable wx app creation because it can cause bugs and they will be warned about it in documentation.

**Comment by set-soft** [comment]
Ok.

I'll suggest you to use some kind of automatic tests like I'm using in my fork.
Your lack of test cases makes you workaround things that we doesn't even know if exist.

**Comment by qu1ck** [comment]
I found the post I had with simple repro. 
https://forum.kicad.info/t/interactive-html-bom-plugin-for-kicad-5-0/11713/97?u=qu1ck

It doesn't work anymore on ubuntu20 with latest kicad from ppa which is good because it means at least some bugs were fixed in wxgtk 3.0.4 or maybe switch to python3 affected it or newer swig.

But either way this plugin must support all systems that kicad supports which includes ubuntu 14 with python2.

I thought about tests but:
1. There is a lot of much more complex code that needs actual tests in this repository, but I have no clue how to do frontend tests yet. Maybe when I have time I'll figure it out.
2. I do have a suite of smoke tests that I run locally. They don't cover edge cases like this, only basic scenarios. Again, maybe I'll port them to a pipeline some day.
3. In this case where issue appears only on old systems which are not available in CI images tests won't even help.

**Comment by qu1ck** [comment]
Damn, should have tested this properly first. Have you done a test run?

```shell
quick@quick-VirtualBox:~/.kicad/InteractiveHtmlBom/InteractiveHtmlBom$ INTERACTIVE_HTML_BOM_NO_DISPLAY=1 ./generate_interactive_bom.py /home/quick/pic_programmer/pic_programmer.kicad_pcb --no-browser
Traceback (most recent call last):
  File "./generate_interactive_bom.py", line 41, in <module>
    config = Config(version)
  File "/home/quick/.kicad/InteractiveHtmlBom/InteractiveHtmlBom/core/config.py", line 93, in __init__
    f = FileConfig(localFilename=self.config_file)
wx._core.PyNoAppError: The wx.App object must be created first!
```

**Comment by qu1ck** [comment]
Fix pushed. I'll update the wiki to document this variable.

Thank you for your contributions.

**Comment by set-soft** [comment]
> But either way this plugin must support all systems that kicad supports which includes ubuntu 14 with python2.

Which KiCad? 5.1.x can use python 2?

> 3. In this case where issue appears only on old systems which are not available in CI images tests won't even help.

Official docker images for Ubuntu 14.04 are available at docker hub:

https://hub.docker.com/_/ubuntu

> Edit: it works if there is no config.ini. I'll add a fix to skip loading it if variable is set.

Yes, I don't have config.ini files.
The tests I'm doing are really naive and oriented to test if we can load UTF-8 boards. Only 36% of the code is covered! And you can find if it was tested just looking at GitHub report about it ;-)

**Comment by qu1ck** [comment]
> Which KiCad? 5.1.x can use python 2?
Any. Both windows and mac are still on python2 and linux switched only recently (since ubuntu 18, debian 10).

> The tests I'm doing are really naive and oriented to test if we can load UTF-8 boards. Only 36% of the code is covered!

That number is meaningless when the only thing checked is that output file exists. 99% of the 36% of code that ran may as well be broken.

I'm not arguing against tests, of course. But good tests take a lot of effort that at this point I can't put in.
